packages/frontend/src/components/DataViz/store/cartesianChartBaseSlice.ts

```diff
@@ -38,8 +59,79 @@ const initialState: CartesianChartState = {
         pivotLayoutOptions: [],
     },
     errors: undefined,
+    chartData: undefined,
+    chartDataLoading: false,
+    chartDataError: undefined,
 };
 
+export const fetchPivotChartData = createAsyncThunk<
+    PivotChartData | undefined,
+    {
+        vizDataModel:
+            | TableDataModel
+            | PieChartDataModel
+            | CartesianChartDataModel;
+        limit: number;
+        sql: string;
+    }
+>(
+    'cartesianChartBaseConfig/fetchPivotChartData',
+    async ({ vizDataModel, limit, sql }) => {
+        const chartData = await vizDataModel.getPivotedChartData({
+            limit,
+            sql,
+            sortBy: [],
+            filters: [],
+        });
+        return chartData;
+    },
+);
+
+export const prepareAndFetchChartData = createAsyncThunk(
+    'cartesianChartBaseConfig/prepareAndFetchChartData',
+    async (_, { getState, dispatch }) => {
+        const state = getState() as RootState;
+
+        const resultsRunner = selectSqlRunnerResultsRunner(state);
+
+        const { selectedChartType, limit, sql } = state.sqlRunner;
+
+        const config = selectChartFieldConfigByKind(state, selectedChartType);
+
+        if (!resultsRunner) {
+            throw new Error('No results runner available');
+        }
+
+        const vizDataModel = getChartDataModel(
+            resultsRunner,
+            config,
+            selectedChartType ?? ChartKind.VERTICAL_BAR,
+        );
+
+        const chartData = await dispatch(
+            fetchPivotChartData({ vizDataModel, limit, sql }),
+        ).unwrap();
+
+        const getChartSpec = (orgColors: string[]) => {
```

Comment:
this allows each chart to trigger a spec retrieval when the current display changes.